repository/src/receive.rs

```diff
@@ -0,0 +1,165 @@
+pub async fn receive<T: RawRepository>(
+    this: &mut DistributedRepository<T>,
+    tip_commit_hash: CommitHash,
+) -> Result<Result<(), String>, Error> {
+    let last_finalized_header = this.get_last_finalized_block_header().await?;
+    let last_finalized_commit_hash = this.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+    let reserved_state = this.get_reserved_state().await?;
+    let mut csv =
+        CommitSequenceVerifier::new(last_finalized_header.clone(), reserved_state.clone())
+            .map_err(|e| {
+                IntegrityError::new(format!("finalized branch is not accepted by CSV: {e}"))
+            })?;
+
+    if this
+        .raw
+        .find_merge_base(last_finalized_commit_hash, tip_commit_hash)
+        .await?
+        != last_finalized_commit_hash
+    {
+        return Ok(Err(
+            "The commit is not a descendant of the last finalized block.".to_owned(),
+        ));
+    }
+
+    // If the branch ends with a fp commmit
+    if let Ok(last_finalization_proof) =
+        format::fp_from_semantic_commit(this.raw.read_semantic_commit(tip_commit_hash).await?)
+    {
+        // Now we consider the direct parent as the received commit
+        // (fp is not treated in the CSV)
+        let commit_hash = this.raw.list_ancestors(tip_commit_hash, Some(1)).await?[0];
+        if commit_hash == last_finalized_commit_hash {
+            return Ok(Err(
+                "The commit is already the last finalized block.".to_owned()
+            ));
+        }
+
+        // Read the commits in the branch and verify them
+        let commits = match read_commits(this, last_finalized_commit_hash, commit_hash).await {
+            Ok(x) => x,
+            Err(CommitError::Commit(error, commit)) => {
+                return Ok(Err(format!("failed to parse commit {commit}: {error}")));
+            }
+            Err(e) => return Err(e.into()),
+        };
+        for (commit, commit_hash) in &commits {
+            if let Err(e) = csv.apply_commit(commit) {
+                return Ok(Err(format!(
+                    "commit sequence verification failed: {e} at {commit_hash}",
+                )));
+            }
+        }
+
+        let (last_commit, last_commit_hash) = commits
+            .last()
+            .expect("already checked that the commit is not same as the last finalized block");
+        if let Commit::Block(_) = last_commit {
+            if csv
+                .verify_last_header_finalization(&last_finalization_proof.proof)
+                .is_err()
+            {
+                return Ok(Err(
+                    "finalization proof is invalid for the last block.".to_owned()
+                ));
+            }
+            advance_finalized(this, *last_commit_hash, last_finalization_proof).await?;
+        } else {
+            return Ok(Err("fp commit must be on top of a block commit.".to_owned()));
+        }
+    // If the branch ends with a block, agenda, or agenda proof commit
```

Comment:
Is this the right place for the comment of the below else block?

Reply:
Below? It's above.